src/rendering/renderers/shared/GCSystem.ts

```diff
@@ -0,0 +1,462 @@
+     */
+    private _updateInstructionGCTick(renderGroup: RenderGroup, gcTick: number): void
+    {
+        renderGroup.instructionSet.gcTick = gcTick;
+
+        for (const child of renderGroup.renderGroupChildren)
+        {
+            this._updateInstructionGCTick(child, gcTick);
+        }
+    }
+
+    /**
+     * Registers a resource for garbage collection tracking.
+     * @param resource - The resource to track
+     * @param type - The type of resource to track
+     */
+    public addResource(resource: GCableEventEmitter, type: GCData['type']): void
+    {
+        // Already being tracked
+        if (resource._gcData)
+        {
+            this.touch(resource);
+
+            return;
+        }
+
+        const index = this._managedResources.length;
+
+        resource._gcData = {
+            lastUsed: this._now ?? performance.now(),
+            index,
+            type,
+        };
+        resource.once('unload', this.removeResource, this);
+
+        this._managedResources.push(resource);
+    }
+
+    /**
+     * Removes a resource from garbage collection tracking.
+     * Call this when manually destroying a resource.
+     * @param resource - The resource to stop tracking
+     */
+    public removeResource(resource: GCable): void
+    {
+        const gcData = resource._gcData;
+
+        if (!gcData) return;
+
+        const index = gcData.index;
+        const last = this._managedResources.length - 1;
+
+        // Swap with last element for O(1) removal
+        if (index !== last)
+        {
+            const lastResource = this._managedResources[last];
+
+            this._managedResources[index] = lastResource;
+            lastResource._gcData.index = index;
+        }
+
+        this._managedResources.length--;
+        resource._gcData = null;
+    }
+
+    /**
+     * Registers a hash-based resource collection for garbage collection tracking.
+     * Resources in the hash will be automatically tracked and cleaned up when unused.
+     * @param context - The object containing the hash property
+     * @param hash - The property name on context that holds the resource hash
+     * @param type - The type of resources in the hash ('resource' or 'renderable')
+     * @param priority - Processing priority (lower values are processed first)
+     */
+    public addResourceHash(context: any, hash: string, type: GCData['type'], priority: number = 0): void
+    {
+        this._managedResourceHashes.push({
+            context,
+            hash,
+            type,
+            priority,
+        });
+
+        this._managedResourceHashes.sort((a, b) => a.priority - b.priority);
+    }
+
+    /**
+     * Marks a resource as recently used, resetting its idle timer.
+     * @param resource - The resource to touch
+     */
+    public touch(resource: GCable): void
+    {
+        if (resource._gcData)
+        {
+            resource._gcData.lastUsed = this._now ?? performance.now();
```

Comment:
can we remove performance.now() as this._now should always be the value we want?